content/blog/tune-parallel/index.Rmd

```diff
@@ -0,0 +1,181 @@
+
+```{r grid-logging-rs, echo = FALSE, message = FALSE, fig.width=5, fig.height=4, out.width = "70%", dev = "svg"}
+load("resamples_times.RData")
+resamples_times %>%
+  dplyr::rename(operation = label) %>% 
+  ggplot(aes(y = id_alt, x = duration, fill = operation)) +
+  geom_bar(stat = "identity", col = "black") +
+  labs(y = NULL, x = "Elapsed Time") + 
+  scale_fill_brewer(palette = "Set2") +
+  theme(legend.position = "top")
+```
+
+The code assigns each of the five resamples to their own worker process which, in this case, is a core on a single desktop machine. That worker conducts the preprocessing then loops over the models. The preprocessing happens once per resample. 
+
+In the new version of tune, there is a control option called `parallel_over`. Setting this to a value of `"resamples"` will select this scheme to parallelize the computations. 
+
+
+## Parallelizing everything
+
+Another option that we can pursue is to take the two loops shown above and merge them into a single loop. 
+
+```{r eval = FALSE}
+all_tasks <- crossing(resamples, configurations)
+
+for (iter in all_tasks) {                           
+   # Create analysis and assessment sets for {iter}
+   # Preprocess data (e.g. formula or recipe)
+   # Fit model {iter} to the {iter} analysis set
+   # Predict the {iter} assessment set
+}
+```
+
+With seven models and five resamples there are a total of 35 separate tasks that can be given to the worker processes. Foe this example, that would allow up to 35 cores/machines to run simultaneously. If we had a validation step, this would also enable the model loop to run in parallel. 
+
+The downside to this approach is that the preprocessing is unnecessarily repeated multiple times (depending on how tasks are allocated to the worker processes). 
+
+Taking our previous example, here is what the allocations look like if the 35 tasks are run across 10 cores: 
+
+```{r grid-logging-all, echo = FALSE, message = FALSE, fig.height = 7, dev = "svg"}
+load("everything_times.RData")
+everything_times %>%
+  dplyr::rename(operation = label) %>% 
+  ggplot(aes(y = id, x = duration, fill = operation)) +
+  geom_bar(stat = "identity", col = "black") +
+  facet_wrap(~ pid, nrow = 2) + 
+  labs(y = NULL, x = "Elapsed Time") + 
+  scale_fill_brewer(palette = "Set2") +
+  theme(legend.position = "top")
+```
+
+For each resample, the preprocessing is needlessly run six additional times. If the preprocessing is fast, this might be the best approach. 
+
+To enable this approach, the control option is set to `parallel_over = "everything"`. 
+
+The default for `parallel_over` is `NULL`. This allows us to check and see if there are multiple resamples. If that is the case, it uses a value of `"resamples"`; otherwise, `"everything"` is used. 
```

Comment:
I sort of feel that this should be pulled out into its own subsection "Automatic strategy detection" or something similar. It seems like its own thing, even if it is just a few sentences